### ISSUE TITLE
FirestoreApi.projects.databases.documents.runQuery throws Exception when query returns list

### ISSUE THREAD

**Issue by pcomans** [dart-lang/googleapis#25]
Running the following code
```
  responses = await (new FirestoreApi(http_client))
      .projects
      .databases
      .documents
      .runQuery(
        //This should really be a RunQueryRequest, but you get the idea
        "{structuredQuery: {from: [{collectionId: messages}], limit: 10, orderBy: [{direction: DESCENDING, field: {fieldPath: createdAt}}]}}",
        "projects/REDACTED/databases/(default)/documents/channels/-KyrWc1BnJF2OlXLhRVV",
        $fields: "document(fields,name)",
      );
```
Throws an exception:
```
type 'List' is not a subtype of type 'Map' of '_json@710331970' where
  List is from dart:core
  Map is from dart:core
```

The reason is that `RunQueryResponse.fromJson` expects a single document (core.Map), while the query above returns a collection (core.List).

I think I am using the query API correctly. Are the API specs incorrect in this case?

**Comment by kentcb** [comment]
I came here to report the same thing. Basically, `RunQueryResponse` assumes there is only a single document returned by the query, so it assumes a `Map` but receives a `List`:

![image](https://user-images.githubusercontent.com/1901832/53680992-8921e280-3d2e-11e9-9481-b93682b47b68.png)

Did you ever find a solution @pcomans?

Can anyone from Google comment on this (very old) issue?

**Comment by kentcb** [comment]
In case it's helpful, here's my repro code:

```dart
  final beginTransactionResponse =
      await firestore.projects.databases.documents.beginTransaction(BeginTransactionRequest(), "projects/redacted/databases/(default)");
  final transactionId = beginTransactionResponse.transaction;

  final query = await firestore.projects.databases.documents.runQuery(
      RunQueryRequest()
        ..transaction = transactionId
        ..structuredQuery = (StructuredQuery()..from = <CollectionSelector>[CollectionSelector()..collectionId = "myCollection"]),
      "projects/redacted/databases/(default)/documents");
```

**Comment by jonasfj** [comment]
This sounds like we're mis-interpreting the discovery document in the code generator, which wouldn't be unlikely... But if the REST API is supposed to be returning a list when it should be returning a map that's is concerning :)
https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.documents/runQuery

If anyone wants to build a minimal example of the REST API doing the wrong thing... I'd be happy to take it up with firestore team. But I suspect the REST API does what it's documented to do.

Instead I suspect that the code we generate is interpreting the response values incorrectly...

---
If someone wants to build a minimal self-contained example demonstrating the issue, I'll try to find time to dig in..

**Comment by kentcb** [comment]
@jonasfj Here's a repro:

```dart
import 'dart:io';

import 'package:googleapis/firestore/v1.dart';
import 'package:googleapis_auth/auth_io.dart';

// TODO: enter your project ID below and put your service credentials into the credentials.json file

final _projectId = "YOUR_PROJECT_ID";
final _databaseId = "(default)";

main(List<String> arguments) async {
  final credentialsFile = File("./credentials.json");
  final credentialsJson = await credentialsFile.readAsString();
  final credentials = new ServiceAccountCredentials.fromJson(credentialsJson);
  final httpClient = await clientViaServiceAccount(credentials, <String>[FirestoreApi.DatastoreScope]);

  final firestore = FirestoreApi(httpClient);

  await firestore.projects.databases.documents.createDocument(Document(), "projects/$_projectId/databases/$_databaseId/documents", "repro");

  // This throws an exception: _TypeError (type 'List<dynamic>' is not a subtype of type 'Map<dynamic, dynamic>')
  await firestore.projects.databases.documents.runQuery(
      RunQueryRequest()..structuredQuery = (StructuredQuery()..from = <CollectionSelector>[CollectionSelector()..collectionId = "repro"]),
      "projects/$_projectId/databases/(default)/documents");
}
```

I've also attached a zip of the project [here](https://github.com/dart-lang/googleapis/files/3078291/repro.zip).

**Comment by jonasfj** [comment]
The discovery API reference is documented in:
https://developers.google.com/discovery/v1/reference/apis#resource

This makes me suspect that there is something in the code generation for this that we're doing wrong. If someone is willing to investigate and patch this I'd be happy to do a review. Otherwise, I'm not sure when I'll have time to go digging in this rabbit hole.

---
it's also possible that this is a bug in the REST API (or the discovery API reference description of it), if anyone can prove this with a REST-only example against the REST APIs:
https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.documents/runQuery

That might be something we can take to the firestore team. Granted it seems far more likely that the codegen in this package does something wrong.

**Comment by izolate** [comment]
Just ran into this bug myself. I believe the [documentation](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases.documents/runQuery#response-body) on the API response body is incorrect. It doesn't make any sense for this endpoint to return a singleton as it is effectively a `list()`/`findAll()` type procedure.

During my tests, this endpoint always returned a list no matter the amount of documents returned.

For example, imagine a hypothetical restaurant app with only a single document in the `restaurants` collection with the property `cuisine: "asian"`:

### Request
```http
POST /v1/projects/$PROJECT_ID/databases/(default)/documents:runQuery HTTP/1.1                                                                                               
Authorization: Bearer $TOKEN                                                                                                                                                             
Content-Type: application/json                                                                                                                                               
Host: firestore.googleapis.com                                                                                                                                               
                                                                                                                                                                             
{                                                                                                                                                                            
    "structuredQuery": {                                                                                                                                                     
        "from": [                                                                                                                                                            
            {                                                                                                                                                                
                "collectionId": "restaurants"                                                                                                                                   
            }                                                                                                                                                                
        ],                                                                                                                                                                   
        "where": {                                                                                                                                                           
            "fieldFilter": {                                                                                                                                                 
                "field": {                                                                                                                                                   
                    "fieldPath": "cuisine"                                                                                                                                       
                },                                                                                                                                                           
                "op": "EQUAL",                                                                                                                                               
                "value": {                                                                                                                                                   
                    "stringValue": "asian"                                                                                                                                     
                }                                                                                                                                                            
            }                                                                                                                                                                
        }                                                                                                                                                                    
    }                                                                                                                                                                        
}
```
### Response
```http
HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8

[
    {
        "document": {
            "createTime": "2019-08-03T17:27:33.383252Z", 
            "fields": {
                "cuisine": {
                    "stringValue": "asian"
                }, 
            }, 
            "name": "projects/$PROJECT_ID/databases/(default)/documents/restaurants/$DOCUMENT_ID", 
            "updateTime": "2019-08-04T05:48:26.236480Z"
        }, 
        "readTime": "2019-08-04T06:58:06.495148Z"
    }
]
```

I believe [this section](https://github.com/dart-lang/googleapis/blob/master/discovery/googleapis/firestore__v1.json#L1788) of the codebase needs to be updated to reflect the reality of the Firestone API response.

@jonasfj Are you able to verify this with the Firestone team? I'd be happy to open up a PR to commit this change once you have done so. Cheers!

**Comment by jonasfj** [comment]
> I believe [this section](https://github.com/dart-lang/googleapis/blob/master/discovery/googleapis/firestore__v1.json#L1788) of the codebase needs to be updated to reflect the reality of the Firestone API response.

Said, section is downloaded from: https://firestore.googleapis.com/$discovery/rest?version=v1

I'll report this to the firestore team.

**Comment by jonasfj** [comment]
I've reported this to the Firestore team, we can't fix it in this project, if the discovery API documents are incorrect we must wait for them to be fixed upstream before we can re-generate the code in this project.

**Comment by frankpepermans** [comment]
This is still an issue in 0.54.0

Any chance this was forgotten? The original issue dates back to 2017

**Comment by jonasfj** [comment]
Closed "won't fix" because the issue is in the underlying API discovery documents.
Feel free to check if the issue have been fixed there, then we can regenerate this API.

**Comment by izolate** [comment]
@jonasfj Did you get a chance to report this to the Firestore team? Would be good to get this on their radar.

**Comment by kentcb** [comment]
@jonasfj Could you please add a link to the issue you reported to the Firestore team?

**Comment by jonasfj** [comment]
I reported it internally.

**Comment by JSANL** [comment]
So we basically just have to wait right now, correct?

**Comment by jonasfj** [comment]
@JSANL, or use the JSON API directly instead of these generated wrappers.

**Comment by SwissCheese5** [comment]
Well this is certainly an issue I didn't expect to hit...

**Comment by Jonas-Sander** [comment]
As this is still open here's some code you can use (its not pretty as I just copied it out of another project):

```dart
import 'dart:convert';

// ignore: import_of_legacy_library_into_null_safe
import 'package:googleapis/firestore/v1.dart';
// ignore: import_of_legacy_library_into_null_safe
import 'package:http/http.dart' as http;

extension FirestoreRunQueryFixedExtension
    on ProjectsDatabasesDocumentsResourceApi {
  /// Runs a query.
  ///
  /// [request] - The metadata request object.
  ///
  /// Request parameters:
  ///
  /// [parent] - The parent resource name. In the format:
  /// `projects/{project_id}/databases/{database_id}/documents` or
  /// `projects/{project_id}/databases/{database_id}/documents/{document_path}`.
  /// For example:
  /// `projects/my-project/databases/my-database/documents` or
  /// `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
  /// Value must have pattern
  /// "^projects/[^/]+/databases/[^/]+/documents/[^/]+/.+$".
  ///
  /// [$fields] - Selector specifying which fields to include in a partial
  /// response.
  ///
  /// Completes with a [RunQueryResponse].
  ///
  /// Completes with a [commons.ApiRequestError] if the API endpoint returned an
  /// error.
  ///
  /// If the used [http.Client] completes with an error when making a REST call,
  /// this method will complete with the same error.
  Future<List<Document>> runQueryFixed(RunQueryRequest request,
      {required http.Client client, required String? parent}) async {
    final projectId = 'YOUR-FIREBASE-PROJECT-ID';
    final urlParentAddition = parent != null ? '/$parent' : '';
    final url =
        'https://firestore.googleapis.com/v1/projects/$projectId/databases/(default)/documents$urlParentAddition:runQuery';
    final body = json.encode(request.toJson());
    final response = await client.post(url, body: body);
    final resBody = response.body;
    final List<dynamic> decoded = json.decode(resBody) as List;
    if (decoded[0]['error'] != null) {
      throw Exception('Firestore Error: $resBody!');
    }
    final docs = [
      for (final docJson in decoded)
        if (docJson['document'] != null)
          Document.fromJson(docJson['document'] as Map)
    ];
    return docs;
  }
}
```

If you query a top level collection you should not have to pass the parent, but will have to select the collection like so 
```dart
  final structuredQuery = StructuredQuery()
    ..where = filter
    ..from = [CollectionSelector()..collectionId = 'YOUR-COLLECTION-ID'];
```

Else you'll have to pass a parent and a collction-ID I think (not sure right now, you'll have to experiment).

Here is also a [gist](https://gist.github.com/Jonas-Sander/4ad27f11aaeeeccb1ed0fef39777d6b6) with some convenience methods.
I don't guarantee it works, I just pasted it. It should still give you a good understanding though.
